### ISSUE TITLE
Native review for trotha01/elm-google-analytics

### ISSUE THREAD

**Issue by trotha01** [elm-lang/package.elm-lang.org#166]
https://github.com/trotha01/elm-google-analytics

I understand that Native reviews are on hold and waiting on 0.17.0
The reason I'm posting is I believe the original reason for the reviews was to see the different use cases for using Native, so this is just to notify of one use case.

This library is a wrapper around the google analytics js function. The js function is small, ~10 lines of code, but I don't believe the functionality it does can be written in Elm currently. 

It adds a script tag with src 'https://www.google-analytics.com/analytics.js' to the dom (which can be done in elm), but it also modifies the window object (which can't be done in elm) for analytics.js to use.

The original javascript function can be found at the top of the page here:
https://developers.google.com/analytics/devguides/collection/analyticsjs/#the_javascript_tracking_snippet

**Comment by evancz** [comment]
I am moving towards the plan outlined [here](https://groups.google.com/forum/#!searchin/elm-dev/native/elm-dev/1JW6wknkDIo/H9ZnS71BCAAJ) and discussed a bit further in [the 0.17 announcement](http://elm-lang.org/blog/farewell-to-frp#what-is-next-).

So I get that this is simple, but I don't think this is the right direction for Elm.

**Comment by trotha01** [comment]
Thanks, I'll check out those links and update

**Comment by domenkozar** [comment]
@trotha01 can you post updates as you go? This is a snippet I'd use on any application, so I'm willing to help out what's needed to make this smooth.

**Comment by trotha01** [comment]
Hi @domenkozar, I would love help with this. Right now I'm trying to figure out how to structure it.

Evan seems to want to move towards a port module instead of using native, but from what I understand (and from the docs) ports are not allowed in libraries. Even if they were allowed in libraries, elm-reactor doesn't support ports (as far as I can tell). Additionally, to compile an elm module with ports, we need to use `--output something.js`. So all users of this library wouldn't be able to use elm-reactor, and would need to compile with the --output flag as well.

If you have ideas on how to structure this to make a nice user experience, let me know. I may try reaching out to the community in the slack channel as well. They may have ideas there.

**Comment by trotha01** [comment]
@evancz Since I can't use Native, and since I can't use ports in libraries, I am not sure how to move forward. How do you recommend making elm wrappers for js libraries?

Use cases include:
- This google analytics snippet
- An Elm library for [webaudio](https://www.w3.org/TR/webaudio/)
- An Elm library for any of the [js apis](https://www.w3.org/standards/techs/js#w3c_all)

Elm seems to severely limit libraries by not allowing use of native js web functionality. Am I missing something here?

**Comment by trotha01** [comment]
After a discussion in the slack #native room, it sounds like native is the only way to use javascript in a library. I made updated the code and continued the thread here:

https://github.com/elm-lang/package.elm-lang.org/issues/176
